### DISCUSSION TITLE
DENIED Redis is running in protected mode

### DISCUSSION THREAD

**Discussion by pratikgithub** [testcontainers/testcontainers-java#4482] [Q&A]
Hi,
I am trying to use the redis image from an private registry.

```
public static GenericContainer<?> redisContainer = new GenericContainer<>(DockerImageName.parse("image_name_with_registry"))
.withExposedPorts(6379);

redisContainer.start();
```

But facing the error :
"redis.clients.jedis.exceptions.JedisDataException: DENIED Redis is running in protected mode because protected mode is enabled, no bind address was specified, no authentication password is requested to clients. In this mode connections are only accepted from the loopback interface. If you want to connect from external computers to Redis you may adopt one of the following solutions: 1) Just disable protected mode sending the command 'CONFIG SET protected-mode no' from the loopback interface by connecting to Redis from the same host the server is running, however MAKE SURE Redis is not publicly accessible from internet if you do so. Use CONFIG REWRITE to make this change permanent. 2) Alternatively you can just disable the protected mode by editing the Redis configuration file, and setting the protected mode option to 'no', and then restarting the server. 3) If you started the server manually just for testing, restart it with the '--protected-mode no' option. 4) Setup a bind address or an authentication password. NOTE: You only need to do one of the above things in order for the server to start accepting connections from the outside."

Could you please help?

**Comment by rameshreddy-adutla** [comment]
The `DENIED Redis is running in protected mode` error happens because Redis's protected mode blocks connections from non-loopback addresses when no password is set. In a Docker/Testcontainers context, the container's IP is not `127.0.0.1` from Redis's perspective.

**Fix — disable protected mode** when creating the container:

```java
new GenericContainer<>("redis:7")
    .withExposedPorts(6379)
    .withCommand("redis-server", "--protected-mode", "no")
    .waitingFor(Wait.forListeningPort());
```

Or **set a password** (recommended for closer-to-production testing):

```java
new GenericContainer<>("redis:7")
    .withExposedPorts(6379)
    .withCommand("redis-server", "--requirepass", "testpassword")
    .waitingFor(Wait.forListeningPort());
```

If you're pulling from a private registry, make sure the image still supports these Redis CLI flags. The same approach works with custom images — it's a Redis config issue, not a Testcontainers one.
